### DISCUSSION TITLE
Integrating Solidity compiler to SDKs

### DISCUSSION THREAD

**Discussion by 0xJepsen** [hiero-ledger/hiero-improvement-proposals#162] [Ideas]
When deploying a Smart Contract on hedera, you need to store its bytecode on the hedera file service. From there, that file acts as a constructor to deploy the contract. 

To get the bytecode of a Smart Contract, you have to compile it using a compiler like [remix](https://remix.ethereum.org/) or [solcjs](https://github.com/ethereum/solc-js). 

Rather than asking our developers to write redundant code or copy-pasting from other dev environments. I propose ease of use integration of solidity compilers into the existing SDKs

Current flow
- Write Smart Contract
- Compile Smart Contract with third-party tools
- Read Smart Contract Byte code 
- deploy smart contract byte code to hedera file service
- pass file ID to`ContractCreateTransaction()`

Flow after
- Write Smart Contract
- Compile and store on Hedera file service (maybe in one command)
- pass file ID to `ContractCreateTransaction()`

Let me know your thoughts.

**Comment by 0xJepsen** [comment]
``` js
    const compiledContract = new CompileContract() // This is what is propose and doesn't exist yet
	.setFilePath("test.sol")

    // Store Contact in file service. Different from eth. Transaction size is smaller on hedera for security 
    const mycontract = await new FileCreateTransaction()
        .setContents(buffer)
        .setKeys([PrivateKey.fromString(myPrivateKey)])
        .setContents(compiledContract)
         // The default max fee of 1 HBAR is not enough to make a file ( starts around 1.1 HBAR )
        .setMaxTransactionFee(new Hbar(2)) // 2 HBAR
        .execute(client);
    
    const TransactionReceipt  = await mycontract.getReceipt(client);
    const fileid =  new FileId(TransactionReceipt.fileId);

    console.log("file: " + fileid);
    // Deploy Contract
    const deploy = await new ContractCreateTransaction()
        .setGas(300)
        .setBytecodeFileId(fileid)
        .setAdminKey(PrivateKey.fromString(myPrivateKey))
        .execute(client);

    const receipt = await deploy.getReceipt(client);//Get the new contract 
    const newContractId = receipt.contractId;        
    System.out.println("The new contract ID is " + newContractId);
```

**Reply by wensheng** [reply]
I don't think this is the job of SDK.  For example, in a debug-compile dev loop, I just want to see if the code compiles.  I don't want to put file on HFS, even on testnet, I don't think Hedera like it either, otherwise HFS will be littered with useless files.  Also it costs HBars to use HFS.  Truffle, Hardhat, Brownie are all doing good jobs compile solidity, different users has difference preferences.  It's relatively simple to get bytecode from compiled solidity and put it to HFS.

**Reply by 0xJepsen** [reply]
Yes, but since we are increasing our support for smart contracts, it is required to store the compiled sc on the HFS before deploying it on the contract service. It would certainly make developers' jobs easier.
